flask_securest/authentication_providers/abstract_authentication_provider.py

```diff
@@ -24,9 +32,35 @@ class AbstractAuthenticationProvider(object):
     to return the username (or another uniquely identifying value)
     which will be logged
     """
-
     __metaclass__ = abc.ABCMeta
 
     @abc.abstractmethod
     def authenticate(self, userstore=None):
         raise NotImplementedError
+
+    @staticmethod
+    def retrieve_credentials_from_request():
```

Comment:
this method should be implemented by the ineriting classes and not here. The headers are not generic are are part of the inner logic of each authentication provider. for example, a new authentication provider password2 could look for a header named pasword2, but other authentication providers should not be aware of this header at all. The logic you copied to the abstract is only relevant to the password authentication provider, and should not be available to other classes.